### ISSUE TITLE
Refactor: understand lower GRAY ELG assignment efficiency

### ISSUE THREAD

**Issue by sbailey** [desihub/fiberassign#165]
Comparing tsk_refactor branch with master using a set of 299 tiles covering DR7.1 range 130 < RA < 150 and 0 < dec < 20, I find that tsk_refactor assigns ~5% fewer ELG targets  to GRAY tiles, making up the difference with stdstars and SKY fibers.  The loss of ELGs is not made up by more ELGs assigned during DARK time.  There is overall a loss of ~1% fewer science targets overall.

Track down specific cases starting with ELGs on GRAY tiles to understand this loss of assignment efficiency.

Here is what I ran for master:
```
basedir=/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor
fiberassign \
    --mtl $basedir/mtl.fits 
    --stdstar $basedir/standards.fits \
    --sky $basedir/skies-dr7.1-0.27.0-lite.fits \
    --footprint $basedir/tiles-lite.fits \
    --outdir $SCRATCH/desi/fiberassign/refactor/master
```

and for tsk_refactor branch:
```
basedir=/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor
fba_run --overwrite \
    --targets $basedir/mtl.fits $basedir/skies-dr7.1-0.27.0-lite.fits \
    --footprint $basedir/tiles-lite.fits \
    --dir $SCRATCH/desi/fiberassign/refactor/output

fba_merge_results \
    --targets $basedir/mtl.fits $basedir/skies-dr7.1-0.27.0-lite.fits \
    --dir $basedir/output \
    --out $SCRATCH/desi/fiberassign/refactor/output
```

I cached a copy of the outputs in `/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor/compare/`.

**Comment by forero** [comment]
@sbailey how is it that the `fba_run` guesses the standard stars file?

**Comment by forero** [comment]
I ran these two commands:
`fiberassign --mtl mtl.fits --sky skies-dr7.1-0.27.0-lite.fits --stdstar standards.fits --outdir output_new/` (refactor version)

`oldfiberassign --mtl mtl.fits --sky skies-dr7.1-0.27.0-lite.fits --stdstar standards.fits --outdir output_old/` (legacy version)

and compared the lenght of the potential tables. These two should match because those are the available targets, there isn't any assignment involved there. However the two differ in almost every tile, some times by a large amount. Here is a histogram using 100 tiles:

![screen shot 2019-01-29 at 11 21 05 am](https://user-images.githubusercontent.com/1848618/51922932-fe109e00-23b7-11e9-8e8e-935b24bdb574.png)

I am puzzled by that difference

**Comment by forero** [comment]
Here is the histogram for all the 588 tiles in that run
![screen shot 2019-01-29 at 11 42 09 am](https://user-images.githubusercontent.com/1848618/51924407-fa324b00-23ba-11e9-8534-04f1e7ee1963.png)

**Comment by forero** [comment]
Here is the same plot using the data provided by @sbailey 
![screen shot 2019-01-29 at 11 54 53 am](https://user-images.githubusercontent.com/1848618/51925297-bccebd00-23bc-11e9-8b21-7d91bfbe344a.png)

**Comment by tskisner** [comment]
I think I understand what is going on.  I ran Stephen's inputs again locally with the new code.  It is instructive to look at the first tile in the sequence (3771) and the last tile in the sequence (45660).  Here is petal number 5 plotted for both.  In this color scheme, gold targets are "plain" standards while green targets are both standards AND science targets (science-only targets are red):

[fiberassign_03771.pdf](https://github.com/desihub/fiberassign/files/2811244/fiberassign_03771.pdf)
[fiberassign_45660.pdf](https://github.com/desihub/fiberassign/files/2811248/fiberassign_45660.pdf)

Important Note:  when considering targets to assign, science targets have a number of remaining observations count.  Standards and sky do not have this count and can be used as much as needed.  

On the first tile in the sequence, we see that 8 dual science/standard objects are assigned, since they typically have high priority.  Then, the remaining 2 standards are forcibly assigned to get 10 per petal (the 10 includes the dual science/standard objects).  Everything is fine.

At the end of the sequence of tiles, some of the science targets have already reached their obs_remaining count, and these are excluded from assignment.  However, the dual science/standard objects count as science objects as well and have infinite obs_remaining, so they basically get assigned during the science assignment phase.  Later, when assigning standards, nothing happens since there are already more than 10 per petal.  These dual science/standards are high priority objects, so will always get assigned ahead of ELGs.

So the question is:  what is the desired behavior for "dual purpose" objects that are both science and standards?  Should the obs remaining be respected for such objects?  If the obs_remaining reaches zero, should that object be excluded from being assigned as a standard or only as a science target?

I believe the old code added the target twice to its internal list in memory (once as a science and once as a standard) and when assigning it as a science target did not count it towards the standards-per-petal count.  That behavior also seems wrong, so we should just decide what exactly we want.

**Comment by forero** [comment]
I ran the two code versions with `--nskypetal 0 --nstarpetal 0`, these new results give the same global 2% difference in the number of assigned ELGs.

![screen shot 2019-01-30 at 11 58 04 am](https://user-images.githubusercontent.com/1848618/51998281-557d3f80-2486-11e9-8532-7ddcc5ceac4d.png)

So, Ted's comment is right (i.e. we have to decide about dual targets) but it seems there is something else driving the difference between the two versions.

**Comment by sbailey** [comment]
In the MTL file, only STD_WD have higher priorities than ELGs, and those are rare so they can't be driving this overassignment.  STD_FAINT and STD_BRIGHT have *lower* priorities than ELGs.  Is the fiberassign code doing something to change the priority of the standards after reading them in?

I think this is the behavior we want:
1. Assign science targets as normal, using NUMOBS_MORE and PRIORITY.
  * most STDs are lower priority than ELGs and thus will only get placed on fibers that don't cover ELG/LRG/QSO/BGS.
2. Count how many assigned science targets also happened to be standards, and count those towards the standards budget.
3. Assign additional standards to unused fibers to try to reach the the standards budget, *not* considering NUMOBS_MORE and PRIORITY.
  * this could assign a STD to an unused fiber that was skipped in (1) because it had already met its NUMOBS_MORE budget, but because we are now ignoring NUMOBS_MORE it gets assigned as a STD.
4. If we still don't have enough standards, start bumping low priority science targets that are not also STD to assign STD instead.
5. Update NUMOBS_MORE based upon what actually got assigned at the end.

**Comment by tskisner** [comment]
While investigating, I noticed that the file skies-dr7.1-0.27.0-lite.fits has SUBPRIORITY set to zero for sky targets.  Because the code uses std::stable_sort, hopefully the order of fibers when assigning sky is reproducible, but sorting a vector based on weights of all zero is not ideal...

**Comment by tskisner** [comment]
Digging deeper...  Looking at the last tile (45660) and fiber 2929 on petal 5.  In the new code, here are the available targets for this fiber:

```
DEBUG: targets avail:  tile 45660, fiber 2929 append ID 30788153270862284 (type=1), total priority 3200.93
NOTE:  DESI_TARGET=1152921504892126473 --> LRG

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 30788153270862407 (type=1), total priority 1500.93
NOTE:  DESI_TARGET=2305843009213693952 --> BGS

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 30788153270862042 (type=3), total priority 1500.05
NOTE:  DESI_TARGET=2305843052163366912 --> BGS + (STD_FAINT | STD_BRIGHT)

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 30788153270862138 (type=1), total priority 1400.85
NOTE:  DESI_TARGET=2305843009213693952 --> BGS

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 30788153270862154 (type=1), total priority 1400.07
NOTE:  DESI_TARGET=2305843009213693952 --> BGS

(The remaining available targets are sky):

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 607248905574285321 (type=4), total priority 0

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 607248905574285332 (type=4), total priority 0

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 607248905574285344 (type=4), total priority 0

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 607248905574285333 (type=4), total priority 0

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 607248905574285345 (type=4), total priority 0

DEBUG: targets avail:  tile 45660, fiber 2929 append ID 607248905574285334 (type=4), total priority 0
```

When the new code gets to assigning this fiber to science:

```
DEBUG: assign unused science: working on tile 45660 at RA/DEC = 139.55 / 17.91
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 30788153270862284, subpriority 0.932502
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 30788153270862407, subpriority 0.930463
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 30788153270862042, subpriority 0.052587
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 30788153270862138, subpriority 0.854437
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 30788153270862154, subpriority 0.0689299
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 607248905574285321 is wrong type (4)
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 607248905574285332 is wrong type (4)
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 607248905574285344 is wrong type (4)
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 607248905574285333 is wrong type (4)
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 607248905574285345 is wrong type (4)
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 available target 607248905574285334 is wrong type (4)
DEBUG: find_best: tile 45660, fiber 2929, target 30788153270862284 science target with no remaining obs
DEBUG: find_best: tile 45660, fiber 2929, target 30788153270862407 science target with no remaining obs
DEBUG: find_best: tile 45660, fiber 2929, target 30788153270862042, type 3 accept with priority = 1500, subpriority = 0.052587, obs_remain = 1
DEBUG: find_best: tile 45660, fiber 2929, target 30788153270862042, type 3 SELECTED
DEBUG: assign unused science: tile 45660, petal 5 fiber 2929 found best object 30788153270862042
```

So it goes through the available targets in priority order.  The highest priority (LRG) has no observations remaining.  Likewise with one of the BGS targets.  Then it gets to the target that is a BGS science target and also a standard, and chooses this.  It does not even examine the lower-priority BGS target with priority class 1400.

Looking at the legacy code, tile 45660, fiber 2929 was assigned to the LRG above.  

Now working backwards, the new code observed this LRG on an earlier tile 11103, fiber 3759:

```
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862284, priority 3200, subpriority 0.932502
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862020, priority 3200, subpriority 0.775135
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862054, priority 3000, subpriority 0.951961
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788147524669387, priority 3000, subpriority 0.680116
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862297, priority 3000, subpriority 0.511196
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862142, priority 3000, subpriority 0.445136
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862166, priority 3000, subpriority 0.228055
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862135, priority 3000, subpriority 0.103369
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862407, priority 1500, subpriority 0.930463
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862138, priority 1400, subpriority 0.854437
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862161, priority 1400, subpriority 0.153926
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 30788153270862154, priority 1400, subpriority 0.0689299
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 607248905574285334 is wrong type (4)
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 607248899828088965 is wrong type (4)
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 607248905574285321 is wrong type (4)
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 available target 607248905574285322 is wrong type (4)
DEBUG: find_best: tile 11103, fiber 3759, target 30788153270862284, type 1 accept with priority = 3200, subpriority = 0.932502, obs_remain = 1
DEBUG: find_best: tile 11103, fiber 3759, target 30788153270862284, type 1 SELECTED
DEBUG: assign unused science: tile 11103, petal 7 fiber 3759 found best object 30788153270862284
```

You can see that the LRG in question has the highest priority (3200.93) of available science targets here.  The legacy code assigned tile 11103, fiber 3759 to target 30788153270862054 (DESI_TARGET=131586) which is the highest priority ELG, but has a lower overall priority.

Perhaps this is a feature of the "improve" step in the legacy code?  I could try commenting that out of the master branch and see if it produces similar results.  That is the only explanation I can think of for why the master branch would choose a lower priority target here.

**Comment by forero** [comment]
I ran into a new issue #172 while trying to use this tutorial https://github.com/desihub/tutorials/blob/master/FiberAssign.ipynb to setup simple mock data to understand the current ELG efficiency issue.

**Comment by forero** [comment]
Using mock targets, as generated in https://github.com/desihub/tutorials/blob/master/FiberAssign.ipynb , the two codes give virtually identical results when the tiles *don't have any overlap*; only when the tiles have some overlap we start having higher numbers in the legacy code.

**Comment by forero** [comment]
There is also an interesting progression with the tile number

In refactor:
```
tileid: 0 - nelg: 3732
tileid: 1 - nelg: 3990
tileid: 2 - nelg: 4105
tileid: 3 - nelg: 4152
tileid: 4 - nelg: 4196
total ELG:  20175
```

In legacy

```
tileid: 0 - nelg: 4432
tileid: 1 - nelg: 4345
tileid: 2 - nelg: 4212
tileid: 3 - nelg: 3984
tileid: 4 - nelg: 3594
total ELG:  20567
```

**Comment by tskisner** [comment]
Indeed, it certainly seems like high-priority targets are being chosen at the beginning by the refactor version.  And it looks like the legacy branch is pushing some of those high priority targets to later tiles in order to get more total ELGs.  This should be made clearer by turning off the "improve" step and comparing the outputs.  Still trying to get to that...

**Comment by sbailey** [comment]
Something is very wrong with the sky fiber assignments on certain gray tiles.  e.g. `/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor/output/tile-26819.fits`:
![sky-26819](https://user-images.githubusercontent.com/218471/52093995-1cba9480-2572-11e9-8a02-6e9ae9487e7d.png)

Code to make the plot to double check my interpretation of SKY fibers:

```python
%pylab
import numpy as np
import fitsio
from desitarget.targetmask import desi_mask

tileid = 26819
columns = ['TARGETID', 'TARGET_RA', 'TARGET_DEC', 'DESI_TARGET', 'PRIORITY', 'SUBPRIORITY', 'FIBER', 'LOCATION']
basedir = '/global/cscratch1/sd/sjbailey/desi/fiberassign/refactor'
fa = fitsio.read(basedir+'/output/tile-{:05d}.fits'.format(tileid), 'FIBERASSIGN', columns=columns)

figure(figsize=(5,5))
for targetnames in ['ELG|LRG|QSO', 'STD_FAINT|STD_BRIGHT', 'SKY']:
    ii = (fa['DESI_TARGET'] & desi_mask.mask(targetnames)) != 0
    plot(fa['TARGET_RA'][ii], fa['TARGET_DEC'][ii], '.', alpha=0.5, label=targetnames)

legend()
```

I picked this tile because compared to master it has 249 *fewer* ELGs assigned but has 542 SKYs assigned (instead of the minimum required 400).

**Comment by sbailey** [comment]
This sky assignment problem isn't unique to gray tiles; see http://portal.nersc.gov/project/desi/users/sjbailey/fiberassign/debug-165/plots/

Also note the non-uniform MWS_ANY targets on tile 44177.

**Comment by tskisner** [comment]
Ok, I'm sure there is some straightforward explanation since such things have never been seen before in lots of testing.  Will try to debug this tonight.

**Comment by sbailey** [comment]
Update: I rechecked the stats with the latest push from @tskisner and the basic issue of low ELG efficiency on GRAY tiles remains.  Summarizing the ratios of refactor/master for various target classes on DARK, GRAY, and BRIGHT layers:
```
ratio of targets new/old
         DARK    GRAY    BRIGHT
ELG      0.989   0.953   ---  
LRG      1.000   0.989   0.972   
QSO      1.002   0.972   0.974   
MWS      1.254   1.263   0.985   
BGS      1.017   1.021   0.995   
STD      0.990   0.870   0.635   
SKY      1.018   1.139   1.126   
```
i.e. overall there are fewer ELG/LRG/QSO targets assigned, fewer STD assigned (but in a separate check it is still assigning the required minimum number of STD), but more SKY and more MWS-nonSTD targets.

**Comment by sbailey** [comment]
Digging into /global/cscratch1/sd/sjbailey/desi/fiberassign/refactor/output/tile-26819.fits:

542 SKY, 839 MWS_ANY, 102 STD - i.e. more skies than are strictly required (400), just over the required number of STD (100), and many MWS_ANY targets that aren't also STD.  I tracked down 146 ELG, LRG, QSO targets that weren't assigned on any tile, but were covered by fibers on this tile that were given to lower priority non-STD targets instead.  Here are the first 25 examples:
```
FIBER  TARGETID           PRIORITY   UNASSIGNED_ID      UNPRIORITY
4329   30787989747535096  2000.621   30787989747535382  3000.026
1457   30788001554499578  2000.015   30788001554499250  3000.288
1700  607248759767695367     0.000   30788001554501104  3000.316
2015   30788019271240820  1400.545   30788025164235106  3000.337
2293   30788025164236364  1500.404   30788025164236125  3000.709
2263  607248777467658306     0.000   30788025164236356  3000.053
1873  607248771562078222     0.000   30788019258655533  3000.434
 687   30787977844097395  2000.493   30787977844097398  3000.515
 651   30787977844099010  1400.137   30787977844098750  3000.329
 756   30787977844102459  1500.285   30787977844101903  3000.349
1574   30788013361467174  1500.544   30788019262849100  3000.027
2216   30788019262853036  1500.507   30788019262853066  3000.552
1213   30787995632142614  1500.727   30787995632143037  3000.065
1209   30787995632143271  2000.667   30787995632144197  3000.579
 961   30787977848291809  1500.880   30787977848291898  3000.007
 788   30787977848295011  1500.990   30787977848295029  3000.712
1533  607248759767695500     0.000   30788013369852329  3000.204
1078  607248753845338167     0.000   30788001541916634  3000.358
1834   30788007460077703  2000.992   30788007460077704  3000.166
 616  607248736086655046     0.000   30787983779038933  3000.873
 637   30787983779040755  1500.034   30787983779040798  3000.674
1281   30787983770651703  2000.039   30787983770651692  3000.665
1717   30788013365659972  2000.069   30788013365659867  3000.594
1907   30788019267043587  1500.546   30788013365661744  3000.659
 796   30787983783231995  1500.690   30787983783231716  3000.153
```
Hopefully debugging the non-assignment history of a few of these UNASSIGNED_ID targets will indicate why they weren't placed.

Guess: when doing the `redistribute_science` step after the initial assignments, it is maybe only checking whether the moved target is *covered* on the other tile, and not whether it is covered by an unassigned fiber?  Or maybe is it noticing that it is covered by a fiber assigned to a lower-priority target, but then moving the target without updating the assignment by bumping the lower priority target?

Another guess: are these all orphans that got bumped from earlier tiles to make room for standards and sky to meet minimum requirements, but then didn't get replaced even though they were higher priority?

**Comment by tskisner** [comment]
Thanks for the additional info.  I will go poke deeper at some of those targets.  A couple notes:

- For the redistribution step (not the "improve" step in the legacy code), science targets are only moved to fibers on other tiles that are empty.  The fibers are iterated in fiber ID order (not based on the current assignment or available targets).  For each fiber, if another tile/fiber is available and can reach the assigned target, and that other tile/fiber is on a petal with fewer science targets, then it moves the target.  Priority is not considered here, just science targets per petal.

- When bumping low priority science targets to place sky / standards, an attempt is made to re-assign bumped science targets to future tile/fibers if there are any available.  If there are none available, then they are "orphaned" as you say.

**Comment by sbailey** [comment]
I've studied the non-assignment of the 25 targets I listed above.

14/25 collided with higher priority targets on neighboring fibers.  Some of these may be recoverable with the "improve" step implemented in master but not tsk_refactor.  It would be helpful to know the change in efficiency when "improve" is turned off in master.

11/25 were originally assigned to a different fiber but then got bumped for a STD or SKY.  But by then the other fibers that covered them had already been assigned to other targets, sometimes lower priority ones.  Currently the tsk_refactor bumping will reassign to an unused fiber, but not to an assigned fiber even if the other assignment is lower priority.  This has the undesirable effect that the presence of low priority targets reduces the efficiency of higher priority targets.  These could be recovered by:
* When a target gets bumped, it can further bump another lower priority target.  Repeat until the bumped target is the lowest one left (though be careful that we don't go full circle and try to bump the SKY or STD that we are originally forcing in).
* Fully assign a tile *and do any necessary SKY and STD bumping* before proceeding with assigning the next tile.  This would reduce the coupling to future tiles and would regain these targets, but potentially lose others because of bypassing the "redistribute" step.

I'll update the previous table with the reason for each target.

**Comment by tskisner** [comment]
I'm working on the master comparison with / without the "improve" step.  With your two bullet points, we currently have a boolean array of all tile / fibers that tracks whether they have already been bumped.  I believe we could use that to control the "recursive bumping".  For the second bullet- that is definitely the more natural thing to do if we are assigning tile-by-tile.  Still TBD if that is ok...

**Comment by sbailey** [comment]
@tskisner showed me in the code where to turn off "improve" for master, and then I re-ran my jupyter numerology:
```
ratio of targets unimproved/improved legacy:
         DARK    GRAY    BRIGHT
ELG      0.988   0.952   ---  
LRG      0.998   1.030   0.947   
QSO      1.002   0.969   0.940   
MWS      1.160   1.276   0.968   
BGS      1.050   1.086   0.988   
STD      1.003   1.044   1.000   
SKY      1.003   1.025   1.190   
```
i.e. the "improve" step significantly helps ELG assignment on the gray layer, which is also where were are missing targets in the refactor branch.
Detail: the LRGs on the GRAY layer are also BGS targets that got flagged as eligible to observe during GRAY time; we haven't disentangled those programs in these tests.

Compared to refactor/legacy (cut-and-pasted from above so you can see them side-by-side):
```
ratio of targets refactor/legacy:
         DARK    GRAY    BRIGHT
ELG      0.989   0.953   ---  
LRG      1.000   0.989   0.972   
QSO      1.002   0.972   0.974   
MWS      1.254   1.263   0.985   
BGS      1.017   1.021   0.995   
STD      0.990   0.870   0.635   
SKY      1.018   1.139   1.126   
```

For those who are watching this ticket and aren't in the weeds, the "improve" step looks for cases like this:
![image](https://user-images.githubusercontent.com/218471/52245904-5f44df80-2898-11e9-8104-9786a01ba8a8.png)
i.e. an unassigned fiber (2) covers a target (B) that is assigned to a fiber (1) that covers an unassigned target (A).  Swapping the assignments around could gain the lost target without losing the original target:
![image](https://user-images.githubusercontent.com/218471/52245916-68ce4780-2898-11e9-8f73-2eeaa09b76f9.png)

**Comment by tskisner** [comment]
Hi @sbailey, I don't think this impacts your tools, but beware that in the outputs from master branch the TARGETS hdu contains incorrect RA/DEC values for at least some sky targets.  The FIBERASSIGN hdu has correct values.  Example:
```
(FIBERASSIGN HDU)
tablist "dr7.1-lite_legacy/tile-26819.fits[1][col TARGETID;TARGET_RA]" | grep 607248736095043610
  20   607248736095043610        135.385269462626
```
and then
```
(TARGETS HDU)
tablist "dr7.1-lite_legacy/tile-26819.fits[3][col TARGETID;RA]" | grep 607248736095043610
  20   607248736095043610        137.116582201887
```
In the TARGETS HDU, many of the sky targets have duplicated RA/DEC.

**Comment by tskisner** [comment]
While investigating the "dual" science / standard target assignment, I think I understand better what is going on.  I'll put this in the context of @sbailey 's desired behavior further up this thread:

>  1. Assign science targets as normal, using NUMOBS_MORE and PRIORITY.  Most STDs are lower priority than ELGs and thus will only get placed on fibers that don't cover ELG/LRG/QSO/BGS.

This is what the code is already doing.  During first pass assignment, dual objects get assigned just like normal science targets.  If there was sufficient density, it might be possible that one could end up with more than 10 per petal (because these are science targets, there is no per-petal threshold).

>  2. Count how many assigned science targets also happened to be standards, and count those towards the standards budget.

The code is currently doing this as well, **BUT** there is a step that happens before this:  science targets are redistributed to later in the survey to load balance the science targets per petal.  If later tiles have empty fibers that can reach dual targets, some of them will be moved.  And this movement of science targets is unaware of the fact that these are also standards.  So in some cases we end up with more than 10 of these dual targets per petal.

When assigning standards, the code sees that there are already more than 10 on a petal and does not assign any additional standards.

>  3. Assign additional standards to unused fibers to try to reach the the standards budget, _not_ considering NUMOBS_MORE and PRIORITY.

The code currently does this, but **does** include the PRIORITY of the standards.  In other words, it uses the priority + subpriority when assigning standards to unused fibers.

In conclusion, when the code is assigning and redistributing science targets, it treats dual targets as just "plain old science targets".  Since it does not consider that these are also standards, we sometimes end up with more than 10 per petal before even considering the assignment of standards.

If this is not the desired behavior, then we should add extra logic when assigning science targets that makes the process aware of the fact that these are also standards, and that it should apply a per-petal threshold on those objects.

Attached is an example of the last tile (45660).  The first case was run with **only** the initial assignment of science targets.  The green targets are "dual" targets.  The code is just treating these as regular science targets.  The second case was run with the initial assignment plus the "redistribution" step.  Note that science targets (including dual targets) were moved to the last tile from earlier in the survey.  The fact that some of these targets are also standards has not been considered in the program flow up to this point.
[fiberassign_45660_science.pdf](https://github.com/desihub/fiberassign/files/2847185/fiberassign_45660_science.pdf)
[fiberassign_45660_science-redist.pdf](https://github.com/desihub/fiberassign/files/2847186/fiberassign_45660_science-redist.pdf)

**Comment by tskisner** [comment]
A test implementation of making the assignment of science targets "aware" of dual targets (and applying a per-petal limit) is now in #174

**Comment by sbailey** [comment]
Trying to close out this mega-issue, I redid the comparison of master with various other versions when assigning DARK+GRAY tiles separately from BRIGHT tiles.  They produce nearly identical assignment efficiencies, except for STD where master overassigns, though I haven't traced exactly where the extra fibers go in the other cases.

master does outperform other options when assigning DARK+GRAY+BRIGHT simultaneously, but this appears to be due to taking advantage of placing LRG+BGS targets on either DARK or GRAY or BRIGHT tiles, which we don't really want to do (we want one DARK obs and one BRIGHT obs).

Summarizing statistics:
```
Dark + Gray tiles:
         m2/m1   r1/m1   r2/m1
  NELG  0.9992  1.0014  1.0034
  NLRG  0.9996  0.9993  0.9992
  NQSO  1.0000  1.0007  1.0007
  NSKY  1.0000  1.0000  1.0000
  NSTD  1.0001  0.9571  0.9539

Bright tiles:
         m2/m1   r1/m1   r2/m1
  NBGS  1.0022  1.0038  1.0016
  NMWS  1.0062  1.0279  1.0086
  NSKY  1.0000  1.0000  1.0000
  NSTD  1.0001  0.9571  0.9539
```
* m2/m1 = master with the "improve" step turned off (m2) compared to master (m1)
* r1/m1 = tsk_refactor branch (r1) compared to master (m1)
* r2/m1 = tsk_refactor assigning tiles serially (r2) compared to master (m1)

Of particular interest is the "r2" case, where I used the tsk_refactor branch but assigned the tiles serially, including bumping targets to make room for SKY and STD to complete a tile before moving on to the next one.  This has nearly the same efficiency as the other methods that mix information across tiles.  Let's study this case in more detail in a separate branch, since it has the nice simplicity of assigning a tile without needing to know about future tiles (which has lower efficiency when mixing DARK/GRAY/BRIGHT, but apparently the same efficiency when isolating them...)

Bottom line: looks good to proceed with merging.
